### ISSUE TITLE
Badly formatted polynomial knowls

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#4394]
On many of the CMF pages (and possibly elsewhere) knowls for large polynomals garble the display.  For example, click on the knowl "Q[x]/(...)" for the cofficient field on the page

http://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/1/118/a/a/

If katex can't handle these large polynomials we should just format them as raw text, and in fact they will probably look better as raw text in situations like this where the coefficients are wider than the knowl (just let the text wrap).

Here is a bold suggestion that wouldn't look as nice but would be much more useful: always display polynomial knowls as raw text, e.g. "x^2 + 3*x + 1", including the *'s so that you can cut and paste them -- there are a lot of situations where it is really annoying not to be able to copy a polynomial you want that you can see clearly right on the LMFDB page.

In fact, I could see an argument to making all polynomials knowls that open with raw text when you click on them, or even better, make it so you can right click on them and copy the raw text to the clip board.

**Comment by jwj61** [comment]
I wonder if the issue is only when the polynomial contains bigint knowls (guessing at the source of the angle brackets).  If the polynomial is already in a knowl, I think you can skip the bigint knowls.

On the sample page, it is a little odd since the polynomial which appears in the knowl is repeated on the next line.  I assume this is because it is not in the nf database and if it were, then the first would be a number field knowl.  In this situation, it might more sense to just write Q[x]/(f) and have the next line say f= ...

That said, I am favor putting the polynomials in raw form in the knowl for the reason you give.  I had tried doing something trickier where you could toggle the display between latex and raw (which I think is better than having the raw version only), but it used javascript and there was a bad interaction with the javascript in the knowl (although in principle, I don't see why that cannot be resolved).

**Comment by AndrewVSutherland** [comment]
@jwj61 I think I like the proposal you made in #2171 even better then having a knowl (it has the virtue pre-selecting the text so all you need to do is click and press ctrl-c to grab the polynomial).  We would have ti see how well this works with very large polynomials, and we would need a way to have some shorthand notation (e.g. $f(x)$) for polynomials that are too big to display in a reasonable way and are not worth even trying to render in katex which if you click on would show in raw form that you could cut and paste).  If we had this I think we could dispense entirely with using bigint knowls for polynomial coefficients.

I think I have come around to your point of view that there just isn't any benefit to ever using bigint knowls for polynomial coefficients, we should only use them for integers and just have two types of polynomials, those we display in full and render in katex (which if you click on them will let you cut and paste the raw polynomial), and those we don't try to display at all but render with a symbol you can click to get the raw polynomial.

**Comment by edgarcosta** [comment]
Creating a generic function `def raw_knowl(display, content)` and its javascript enable, that does a similar job to the bigint knowls and allows for easy copying of content should be easy.
Then we can have various functions that use the generic one for displaying polynomials, integers, etc...

If we are going via this approach I have a couple of questions:
- do we want to use some different styling for these knowls? I think we might want something a bit more subtle.
Otherwise, pages like [1], will become overwhelming with blue underlines.

- how we will decide when to use it and what to display? For single variable polynomials and integers, this should be easier.
But things get a bit more tricky for multivariable polynomials or q-expansions.
For example, take a look at [1]. Is there any bigint knowl we would want to keep? 
- should we put a limit on the box with the raw content? 
For example, for numbers, do we want to see [156888240371914..]📋 or the number taking over many lines?

[1]: http://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/1/118/a/a/

**Comment by AndrewVSutherland** [comment]
@edgarcosta Here are my thoughts on your questions (but others should feel free to chime in):

- I agree 100% about the link styling, in fact I find the blue underlines annoying already and would be in favor of removing them everywhere with a global CSS change (a.link text-decoration:none).  I note that there are no underlines on our front page (https://www.lmfdb.org/), if it's blue, it's probably a link, and if you are not sure, you can always hover or click.
- Regarding the [CMF page](http://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/1/118/a/a/) you asked about, I would get rid of every bigint on that page, I would make all the polynomials raw_knowls, including the q-expansion, where I might use say b1, b2, ... as replacements for \beta_1, \beta_2.  The user is probably better off downloading the q-expansion in this case, since that will create the number field and all the necessary polynomial rings for them, but I think it is still be useful to be able to just reach in and grab the raw text -- maybe I just want to grab a_2).  The only question left is what to use as a "thumbnail" for the polynomials that are too big to display in full.  Showing just the first one or two coefficients (or how many easily fit) and then \cdots might be the way to go.
-  I'm not sure I completely understand your last question, but I think your asking whether to wrap text inside the raw box or not.  I think my preference would probably be to wrap for display purposes but not insert any carraige returns or backslashes, just let me cut and past the whole thing (Sage and Magma are both perfectly happy to let you paste huge polynomials into the command line as single lines of text)..  But I would probably also be fine with one really long line if others preferred that (what I don't like is having backslashes inserted for me and line breaks that bear no relation to the size of the bounding box/knowl).

**Comment by roed314** [comment]
* I like being able to distinguish between knowls, which won't take you away from the page, and links, which will.  That being said, I think that we should not underline polynomials and bigints.
 * I agree with Drew's suggestions about the CMF page, including just using the first few coefficients and then cdots.  This should be tunable: I think it would be kind of annoying to have the q-expansion just say `q + ...` if the coefficient of `q^2` is too large.  On a related note, I assume that if a user clicks on Display 100 coefficients then all 100 coefficients will be displayed, even if that requires a lot of space.

**Comment by AndrewVSutherland** [comment]
@roed314 What if the hover text for knowl links showed the underline?  I think in most cases it's easy to tell from context which links are likely to be knowls and which are not (e.g. captions and terms that you might want to know the definition of are knowls but most other blue links are not).  I think the hover text would give you a quick visual cue, and the harm of occasionally taking an unsuspecting user away from a page and forcing them to hit the back button to go back has to be weighed against the impact of making every page busier than it needs to be (I also think it makes text containing many knowlified words harder to read).

**Comment by edgarcosta** [comment]
@AndrewVSutherland, perhaps my question makes more sense with examples.
Here are a couple of examples of how to display a large number in a copy-friendly way:
https://math.mit.edu/~edgarc/example.html

Obviously, there are many other options, but each one has its advantages and disadvantages.
For example. the first example doesn't take any space and is very convenient if we really believe that the user just wants to copy its content.

**Comment by AndrewVSutherland** [comment]
@edgarcosta I like the middle one with a vertical scroll bar.

**Comment by edgarcosta** [comment]
I added one more example which might be the best compromise.
This example uses `textarea` which allows the user to easily reformat the box

**Comment by AndrewVSutherland** [comment]
@edgarcosta Nice, I like it!  The one thing I might change is to make it possible to select just part of the text (but I would have it start with all the text selected so you only have to type ctrl-C to grab it, and double clicking should always select everything)

**Comment by jwj61** [comment]
I just want to get an idea of the basics of the design here.

For reference, I think K-M show everything in full on the page in typeset form.  If you click on a polynomial, it toggles in place to raw form and selects it.  If you click another polynomial, the first goes back to typeset along with the change just described.  (Since you click on the item to toggle it, it pretty much has to be auto-selected or else users might accidentally toggle it back while trying to select the raw version.)

Another issue I had raised was for typeset polynomials in knowls.  I wanted to add a button in the knowl to toggle it between raw and typeset.  Then the user can select on the raw version and paste elsewhere.  The only problem here was in getting the javascript to work ok with knowl javascript.  If we were applying this across the site, it would be simpler since the script tags could be present in the page headers instead of loading them inside the knowl.

A simplified version of the last paragraph would be for the knowl to simply contain raw form and not have a typeset version.  This satisfies one goal (easy to select) but not another (looks nice for someone who just wants to peek and see the thing).

I get the feeling we are now talking about another variation on this theme, yes?

**Comment by roed314** [comment]
@AndrewVSutherland I'd be open to just adding a dotted underline on hover; I'd like to see what it feels like before committing to it though.

**Comment by AndrewVSutherland** [comment]
@jwj61 The one thing we have to deal with that K-M do not is that we have polynomials that are too big to typeset on the page, and some that are too big to typeset at all (katex just chokes on them).  For any polynomial that is small enough to render on the page reasonably (ideally on one line) I think it should appear typeset and clicking will let you select it -- whether that happens in place or not I don't have a strong opinion about (although doing it in place is kind of cool).

For polynomials that are too large to render on the page, I think we should have some sort of thumbnail (perhaps the first few terms of the polynomial, if they will fit), and then clicking on it would open the raw box Edgar has been playing with.

There is an intermediate size range where your suggestion of providing a "typeset inside knowl" option might make sense, but it sounds more complicated and it won't handle the really big cases.  But if people are motivated to pursue three display options for small/medium/large polynomials rather than the two I had in mind, I wouldn't object.

**Comment by jwj61** [comment]
Thanks for the explanation.  If we have code for the K-M version, then it is likely it can work inside a knowl which removes the need for a separate medium solution.

**Comment by AndrewVSutherland** [comment]
@roed314 On blue.lmfdb.xyz you can see what the knowl links look like with the dotted underlines removed but still appearing on hover.  I invite everyone to do a side-by-side comparison with beta.lmfdb.xyz on some of your favorite pages -- I think the pages look less busy and are easier to read without the dotted underlines.

Edit: I just tested these pages side by side with someone whose vision is slightly less than perfect (she doesn't wear glasses but uses reading glasses on occasion) and the results were striking.  With her glasses off she found that the page with dotted underlines looked blurrier, like they had a fuzzy shadow, which is also my impression on a hi-res screen unless I zoom in.  But with her glasses on she didn't have any problem reading the captions and didn't really have a strong preference either way, other than thinking the screen was less busy without the dotted underlines.  This perhaps explains the difference in our perceptions of the pages @roed314 -- I expect your eyesight is better than mine!

**Comment by jwj61** [comment]
They are less busy, but difference between a knowl link and a regular link is pretty subtle if you have to hover to see it.  I am worried that for many users will perceive this as no difference.

**Comment by JohnCremona** [comment]
Surely novice LMFDB users will not expect any links to pop open info on the same page, so the fact that knowls do that would come as a (nice) surprise.  I cannot imagine anyone complaining that they clicked on something expecting it to be a knowl and were infuriated by being taken to a different page!

When there is no line until you hover (as on blue now), I think that attention is likely to be drawn by the little pop-up cpation in a black rectangle which shows the knowl's title and id reference.  Novices may well think that that is *all* there is and not click at all!  Would it be too fussy for the back rectangle to pre-pend some text like "click to find out about ..."?

**Comment by AndrewVSutherland** [comment]
I agree with @JohnCremona, both that users expect to click on blue links, and that the hover text might actually being doing the opposite of what we intend and it would be better to just say "click to learn about X".  Users familiar with the LMFDB (which is most of our visitors) will know what knowls are and I don't think they need the dotted, while users not familiar with the LMFDB probably have no intuition other than links are blue (they have to have clicked on a blue link to have even reached a page with a knowl on it), and after that if they do notice the dotted underline they might well be led to think that just means there is hover text and *not* click on it.

**Comment by JohnCremona** [comment]
OK, with the minor proviso that I sometimes find it helpful that hovering reveals the knowl's id!  But users will not care about the id.

And it's not quite true that you'll only see a page with knowls by clicking on another LMFDB page, since more and more people are putting in links to home pages in their papers!

**Comment by AndrewVSutherland** [comment]
@JohnCremona I hadn't actually noticed the knowl id in the tooltip (probably because I rarely hover long enough to see the tooltip), that is useful!  Maybe we could have it display the knowl id when you are running in debug mode?  That would be useful to developers, who are the main people who need to know the knowl  ids.

**Comment by jwj61** [comment]
I was thinking of someone who has used the LMFDB somewhat to be aware that sometimes you get a knowl and sometimes you get a link.  Such a person might want to know the difference (I often delay clicking on links on other sites so I can finish with the current page).  The question is, will the hover behavior be enough that they quickly discern the difference.  I am not sure I would ever notice.

**Comment by AndrewVSutherland** [comment]
@jwj61 Fair enough, I guess my claim is that a familiar user will know when to expect a knowl (e.g. captions are always knowls) and if you are ever unsure you can always hover (as JC noted, the tool tip is pretty obvious, although it takes longer than the underline to show up).  But it sounds like there is a diversity of opinion on this question so we should think about it carefully before making any changes.  That shouldn't slow down addressing this issue (I would just make the links look like knowl links for now and we can debate what knowl links should look like in parallel)

**Comment by JohnCremona** [comment]
Good idea to only display the knowl id in debug mode: +1.

Reading the last few comments on a phone reminded me that there is no equivalent to hovering on a touch screen (that I know of).  I'm not sure what we can do about that. I do have a permanent LMFDB bookmark on my phone for emergencies.  Ease of use on tablets is certainly something we should support.

**Comment by edgarcosta** [comment]
@jwj61, we have been playing around with your raw mode and how to incorporate it for larger inputs.
One idea is to use `textarea` with some cutoff, which we would need to discuss, but here are some rough examples that demonstrate what I mean by it:
- https://pink.lmfdb.xyz/ArtinRepresentation/9.18913710080000000.18t272.a.a look at the different roots
- https://pink.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1/118/a/a/ see the defining polynomial
The threshold is not perfectly tuned and we could also adjust the `textarea` a posteriori to match the original size of the typeset.
what do you think?

**Comment by jwj61** [comment]
I like the original better.  On pink, I fell like I clicked to see something, but then I can't see it.  In the case of the Artin representation, the raw form takes about the same amount of space as the typeset version, soit isn't really saving screen space.

On the other hand, modular form polynomial is being partially concealed, so presumably it would take up too much space if given in raw form.  Maybe make an optional argument to the function which makes the span so we have the option to request it for information which would get much bigger when put in raw form.

**Comment by edgarcosta** [comment]
I have more progress on this. The CSS for the buttons isn't great, but things have improved:

- https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1/118/a/a/ (compare defining polynomial, know for coeff field and basis)
- https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/8449/2/a/h/ (hecke kernels at the end)

If you are curious to see the changes, the branch is https://github.com/edgarcosta/lmfdb/tree/copy_raw

I would like to fix q-expansions before submitting a PR, but that might take a while as I have a very busy week coming up.

**Comment by AndrewVSutherland** [comment]
The link https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/8449/2/a/h/ currently gives a server error:
```
  File "/home/lmfdb/lmfdb-git-tori/lmfdb/utils/web_display.py", line 426, in compress_polynomial
    monomial = monomial_str.format(n)
IndexError: Replacement index 2 out of range for positional args tuple
```

**Comment by edgarcosta** [comment]
Fixed, `T_{2}` got the formatting confused ;)

**Comment by edgarcosta** [comment]
@AndrewVSutherland, do you have a concrete suggestion of what should do regarding the q-expansion?

Let's use the example https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1/118/a/a/

should  I compress term by term? or just give up and do `q + ....`

**Comment by edgarcosta** [comment]
and for the trace formula? should we just use scientific notation?

**Comment by AndrewVSutherland** [comment]
@edgarcosta I would treat both as polynomials (remove the O(q^n) at the end from the cut-and-pasteable text).  In the case of the Q-expansion this would a polynomial in Z[b1,...,bn][q] (I would use variables b1, b2, ... rather than \beta_1, \beta_2, ... so that one could actually cut and paste this if you have already defined the polynomial ring, or the values of b1,b2,... using the cut-and-pasteable text from the basis entries (in which case they would evaluate to polynomials in Q(a)[q]).

**Comment by roed314** [comment]
When the weight is large, the three things I think are helpful to get out of the coefficients are
* for which `n` is the coefficient of `q^n` zero
* What is the order of magnitude of growth in the coefficient size (under the different embeddings of the field).
* what does the trace form look like

The first two objectives are achieved pretty well by the embedded modular forms lower on the page.

**Comment by roed314** [comment]
@AndrewVSutherland I think Edgar was asking about what should be displayed before switching to the pastable view.

**Comment by AndrewVSutherland** [comment]
There I would just treat them like the other polynomials (first and last terms separated by ... with coefficients also compressed if needed.  For the q-expansion, if the leading or trailing terms have coefficients that are polynomials in the betas that need to be compressed, I would do the same thing.

**Comment by edgarcosta** [comment]
It is still unclear to me what to do.
What would you display for the q-expansion here: https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1/118/a/a/

At the moment we display:
```
q +(4490781853906528 - \beta_{1}) q^{2} +(11\cdots04 + 2251433924 \beta_{1} + \beta_{2}) q^{3} +(75\cdots32 - 40679477561379210 \beta_{1} + 3291802 \beta_{2} + \beta_{3}) q^{4} +(42\cdots50 + 53\cdots23 \beta_{1} - 1520180837375 \beta_{2} + 10132 \beta_{3} + \beta_{4}) q^{5} + ...
```

should this become
```
q + (44 \cdots 28 - \beta_{1}) q^{2}  + (11\cdots04 + 22\cdots24 \beta_{1} + \beta_{2}) q^{3} +(75\cdots32  + \cdots + \beta_{3}) q^{4} +(42\cdots50 + \cdots + \beta_{4}) q^{5} 
```
or simply give up and display
```
q + (44 \cdots 28 - \beta_{1}) q^{2}  + \cdots + O(q^10)
```

**Comment by AndrewVSutherland** [comment]
Of those two, I would say the latter, but what I was suggesting is 

```
q + (44 \cdots 28 - \beta_{1}) q^{2}  + \cdots +  (25\cdots 04\beta_1+\cdots+\beta_8)q^9 +  O(q^10)
```
I.e. don't treat `O(q^10)` as the last term, just treat it as something to be added after the compressed polynomial.

**Comment by edgarcosta** [comment]
tks! that comment is quite helpful

**Comment by roed314** [comment]
It may also be worth doing this for polynomials that arise in search results.  For example, [this search](https://beta.lmfdb.org/NumberField/?hst=List&degree=40&is_galois=no&search_type=List) has a polynomial column that is so wide you don't see that there are other columns to the right.

**Comment by edgarcosta** [comment]
I fully agree, but at the moment I would like to only focus on CMFs, so we tweak things before affecting other areas.

**Comment by edgarcosta** [comment]
Bug report from @roed314:

> I've managed to get the x->x stuck in copyable mode (so it won't switch back to latex display).  Here are steps to reproduce:
> 1. Switch to copyable mode
> 2. Double click in the text area (yes, I know that you can use the copy button, but some people are used to doing this to select text for copying).
> 3. Now you can no longer go back.

**Comment by edgarcosta** [comment]
I have implemented the q-exp.

Here is an example:
https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1/118/a/a/

**Comment by AndrewVSutherland** [comment]
That example looks great!  But there seems to be a problem on https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1496/4/a/f/

**Comment by edgarcosta** [comment]
That has been fixed. :)


Weight 1 is currently broken, but here are some working as intended examples:
- https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/819/2/fm/d/
- https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/54/24/a/a/
- https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/46/10/a/e/

**Comment by AndrewVSutherland** [comment]
These look fabulous!  (insert Chef's kiss emoji here)

**Comment by edgarcosta** [comment]
Weight 1 has been fixed.

Here is an example: https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/1487/1/b/a/

**Comment by edgarcosta** [comment]
You can check more (potentially interesting) random examples via:

https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/?dim=16-20&search_type=Random

**Comment by AndrewVSutherland** [comment]
There is a problem with the latex exponent on the factored Hecke poly for p=3 on https://tori.lmfdb.xyz/ModularForm/GL2/Q/holomorphic/897/2/r/f/

**Comment by edgarcosta** [comment]
It has been fixed
